google/cloud/spanner/internal/time.cc

```diff
@@ -0,0 +1,216 @@
+#include <ratio>
+#include <sstream>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+// All the civil-time code assumes the proleptic Gregorian calendar, and
+// 24-hour days divided into 60-minute hours and 60-second minutes.
+
+using nanoseconds = std::chrono::nanoseconds;
+using seconds = std::chrono::seconds;
+using system_clock = std::chrono::system_clock;
+using time_point = system_clock::time_point;
+
+//
+// Duration
+//
+
+google::protobuf::Duration ToProto(nanoseconds ns) {
+  auto nanos = ns.count();
+  google::protobuf::Duration proto;
+  proto.set_seconds(nanos / nanoseconds::period::den);  // rounds toward zero
+  proto.set_nanos(nanos % nanoseconds::period::den);    // (a/b)*b + a%b == a
+  return proto;
+}
+
+void ToProto(nanoseconds ns, google::protobuf::Duration* proto) {
+  *proto = ToProto(ns);
+}
+
+nanoseconds FromProto(google::protobuf::Duration const& proto) {
+  nanoseconds ns(proto.nanos());
+  ns += seconds(proto.seconds());
+  return ns;
+}
+
+//
+// Timestamp
+//
+
+google::protobuf::Timestamp ToProto(time_point tp) {
+  std::time_t t = system_clock::to_time_t(tp);
+  time_point ttp = system_clock::from_time_t(t);
+  auto ss = std::chrono::duration_cast<nanoseconds>(tp - ttp);
+  if (ss.count() < 0) {
+    t -= 1;
+    ss += seconds(1);
+  }
+  google::protobuf::Timestamp proto;
+  proto.set_seconds(t);
+  proto.set_nanos(ss.count());
+  return proto;
+}
+
+void ToProto(time_point tp, google::protobuf::Timestamp* proto) {
+  *proto = ToProto(tp);
+}
+
+time_point FromProto(google::protobuf::Timestamp const& proto) {
+  time_point tp = system_clock::from_time_t(proto.seconds());
+  tp += nanoseconds(proto.nanos());
+  return tp;
+}
+
+namespace {
+
+// A duration capable of holding subsecond values at high precision.
+using femtoseconds = std::chrono::duration<std::int64_t, std::femto>;
+
+// Convert a std::time_t into a Zulu std::tm.
+std::tm ZTime(std::time_t const t) {
+  std::time_t sec = t % (24 * 60 * 60);
+  std::time_t day = t / (24 * 60 * 60);
+  if (sec < 0) {
+    sec += 24 * 60 * 60;
+    day -= 1;
+  }
+  day += 719468;  // since 1970-01-01
+  std::time_t const era = (day >= 0 ? day : day - 146096) / 146097;
+  std::time_t const doe = day - era * 146097;
+  std::time_t const yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
+  std::time_t const y = yoe + era * 400;
+  std::time_t const doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
+  std::time_t const mp = (5 * doy + 2) / 153;
+  std::time_t const d = doy - (153 * mp + 2) / 5 + 1;
+  std::time_t const m = mp + (mp < 10 ? 3 : -9);
+  std::tm tm;
+  tm.tm_year = static_cast<int>(y + (m <= 2 ? 1 : 0) - 1900);
+  tm.tm_mon = m - 1;
+  tm.tm_mday = d;
+  tm.tm_hour = sec / (60 * 60);
+  tm.tm_min = (sec / 60) % 60;
+  tm.tm_sec = sec % 60;
+  return tm;
+}
+
+// Convert a Zulu std::tm into a std::time_t.
+std::time_t TimeZ(std::tm const& tm) {
```

Comment:
Some of the same concerns about magic numbers here.

Reply:
I made it so the comment applies to both `ZTime()` [absolute -> civil] and `TimeZ()` [civil -> absolute].

Aside: `ZTime()` is just like `gmtime()`.  Unfortunately there is no standard inverse function, so we have to write our own, and at that point we're better off writing a `gmtime()` replacement too so that there is some correspondence between them, and so we can avoid `gmtime()` portability issues.